ucp/_libs/utils.pyx

```diff
@@ -42,41 +42,52 @@ def get_buffer_data(buffer, check_writable=False):
     return data_ptr
 
 
-def get_buffer_nbytes(buffer, check_min_size, cuda_support):
+cpdef size_t get_buffer_nbytes(buffer, check_min_size, bint cuda_support):
     """
     Returns the size of the buffer in bytes. Returns ValueError
     if `check_min_size` is greater than the size of the buffer
     """
 
-    iface = None
+    cdef dict iface = None
     if hasattr(buffer, "__cuda_array_interface__"):
         iface = buffer.__cuda_array_interface__
         if not cuda_support:
-            msg = "UCX is not configured with CUDA support, please add " \
-                  "`cuda_copy` and/or `cuda_ipc` to " \
-                  "the UCX_TLS environment variable"
-            raise ValueError(msg)
+            raise ValueError(
+                "UCX is not configured with CUDA support, please add "
+                "`cuda_copy` and/or `cuda_ipc` to "
+                "the UCX_TLS environment variable"
+            )
     elif hasattr(buffer, "__array_interface__"):
         iface = buffer.__array_interface__
 
+    cdef size_t i
+    cdef int64_t s
+    cdef size_t itemsize
+    cdef uint64_t[::1] shape
+    cdef int64_t[::1] strides
+    cdef size_t nbytes
     if iface is not None:
         import numpy
-        itemsize = int(numpy.dtype(iface['typestr']).itemsize)
+        itemsize = numpy.dtype(iface["typestr"]).itemsize
         # Making sure that the elements in shape is integers
-        shape = [int(s) for s in iface['shape']]
-        nbytes = reduce(operator.mul, shape, 1) * itemsize
+        shape = numpy.array(iface["shape"], dtype=numpy.uint64)
```

Comment:
What is `iface["shape"]` here? A tuple? In Numba previously @shwina found that allocating a numpy array to do a cumulative product of a tuple is much more expensive then just doing an `itertools.reduce` for example. We should likely be even more efficient in using the Python C-API here if we wanted and iterating over the tuple via that.

Reply:
Yeah wasn't planning to stick with NumPy. In any event there are a few ways of coercing something like a `tuple` to a C-style array. Of them just using the builtin `array` should be fine for small arrays (as seen below). This was run on a DGX-1.

```python
In [1]: from array import array 
   ...: import struct 
   ...:  
   ...: import numpy                                                            

In [2]: t = (2, 3, 4)                                                           

In [3]: %timeit numpy.array(t)                                                  
1.17 µs ± 2.5 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [4]: %timeit array("l", t)                                                   
243 ns ± 0.105 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [5]: %timeit struct.pack(f"{len(t)}l", *t)                                   
368 ns ± 1.8 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

Reply:
This now uses `array`. So marking this as resolved.